pkg/serviceaccounttoken/secret_test.go

```diff
@@ -38,63 +65,99 @@ func TestEnsureSecretForServiceAccount(t *testing.T) {
 					Name: "test-token-abcde",
 				}},
 			},
-			wantSecret: &v1.Secret{
+			wantSA: defaultWantSA,
+			existingSecret: &v1.Secret{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "test-token-abcde",
 					Namespace: "default",
+					Annotations: map[string]string{
+						"kubernetes.io/service-account.name": "test",
+					},
 				},
 				Data: map[string][]byte{
 					"token": []byte("abcde"),
 				},
+				Type: v1.SecretTypeServiceAccountToken,
 			},
+			wantSecret: defaultWantSecret,
 		},
 		{
-			name: "service account with existing secret returns it",
+			name:    "returns error for nil service account",
+			wantErr: true,
+		},
+		{
+			name: "service account with invalid secret is updated with new secret",
 			sa: &v1.ServiceAccount{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "test",
 					Namespace: "default",
 				},
 				Secrets: []v1.ObjectReference{{
-					Name: "test-token-abcde",
+					Name: "wrong",
 				}},
 			},
-			wantSA: &v1.ServiceAccount{
+			wantSA:     defaultWantSA,
+			wantSecret: defaultWantSecret,
+		},
+		{
+			name: "secret of wrong type gets recreated",
+			sa: &v1.ServiceAccount{
 				ObjectMeta: metav1.ObjectMeta{
 					Name:      "test",
 					Namespace: "default",
 				},
 				Secrets: []v1.ObjectReference{{
-					Name: "test-token-abcde",
+					Name: "test-token-xyz",
 				}},
 			},
+			wantSA: defaultWantSA,
 			existingSecret: &v1.Secret{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:      "test-token-abcde",
+					Name:      "test-token-xyz",
 					Namespace: "default",
+					Annotations: map[string]string{
+						"kubernetes.io/service-account.name": "test",
+					},
 				},
 				Data: map[string][]byte{
 					"token": []byte("abcde"),
 				},
+				Type: v1.SecretTypeOpaque,
 			},
-			wantSecret: &v1.Secret{
+			wantSecret: defaultWantSecret,
+		},
+		{
+			name: "secret for wrong service account type gets recreated",
+			sa: &v1.ServiceAccount{
 				ObjectMeta: metav1.ObjectMeta{
-					Name:      "test-token-abcde",
+					Name:      "test",
 					Namespace: "default",
 				},
+				Secrets: []v1.ObjectReference{{
+					Name: "test-token-xyz",
+				}},
+			},
+			wantSA: defaultWantSA,
+			existingSecret: &v1.Secret{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      "test-token-xyz",
+					Namespace: "default",
+					Annotations: map[string]string{
+						"kubernetes.io/service-account.name": "wrong",
+					},
+				},
 				Data: map[string][]byte{
 					"token": []byte("abcde"),
 				},
+				Type: v1.SecretTypeServiceAccountToken,
 			},
-		},
-		{
-			name:    "returns error for nil service account",
-			wantErr: true,
+			wantSecret: defaultWantSecret,
 		},
 	}
 
 	for _, tt := range tests {
 		t.Run(tt.name, func(t *testing.T) {
+			t.Parallel()
```

Comment:
This will always run the last subtest due to that quirk of Go with iteration value being shared by goroutines, so you'd need to capture the value just for the current iteration.

```
for _, tt := range tests {
    tt := tt
    t.Run(tt.name, func(t *testing.T) {
	t.Parallel()
```

And the test itself can be run in parallel:
```
func TestEnsureSecretForServiceAccount(t *testing.T) {
   t.Parallel()
```

Reply:
Oops, fixed